File: packages/core/src/jws.ts
/**
 * Detached-payload JWS (compact) with Ed25519/EdDSA, per ARKY-TIM-v1 §6 and
 * RFC 7797 (b64:false). The compact form is `<protected>..<signature>` — the
 * payload segment is empty; the payload is the JCS canonical bytes supplied
 * separately when signing/verifying.
 *
 * Clean-room: base64url and the RFC 7797 signing-input construction are
 * implemented here over @noble/curves ed25519 (not via jose).
 */

import { ed25519 } from '@noble/curves/ed25519';

export interface ProtectedHeader {
  alg: 'EdDSA';
  b64: false;
  crit: ['b64'];
  typ?: string;
  kid?: string;
}

export function base64urlEncode(bytes: Uint8Array): string {
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

export function base64urlDecode(s: string): Uint8Array {
  const pad = s.length % 4 === 0 ? '' : '='.repeat(4 - (s.length % 4));
  const bin = atob(s.replace(/-/g, '+').replace(/_/g, '/') + pad);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

/**
 * RFC 7797 signing input for a detached, unencoded payload:
 *   ASCII(BASE64URL(protected)) || '.' || payload_bytes
 */
function signingInput(protectedB64: string, payload: Uint8Array): Uint8Array {
  const prefix = new TextEncoder().encode(protectedB64 + '.');
  const out = new Uint8Array(prefix.length + payload.length);
  out.set(prefix, 0);
  out.set(payload, prefix.length);
  return out;
}

/** Sign canonical `payload` bytes; returns compact `<protected>..<signature>`. */
export function signDetached(payload: Uint8Array, privateKey: Uint8Array, kid?: string): string {
  const header: ProtectedHeader = {
    alg: 'EdDSA',
    b64: false,
    crit: ['b64'],
    typ: 'JWS',
    ...(kid ? { kid } : {}),
  };
  const protectedB64 = base64urlEncode(new TextEncoder().encode(JSON.stringify(header)));
  const sig = ed25519.sign(signingInput(protectedB64, payload), privateKey);
  return `${protectedB64}..${base64urlEncode(sig)}`;
}

/** Decode the protected header of a compact JWS. */
export function decodeProtectedHeader(jws: string): ProtectedHeader & Record<string, unknown> {
  const protectedB64 = jws.split('.')[0];
  return JSON.parse(new TextDecoder().decode(base64urlDecode(protectedB64)));
}

/**
 * Verify a detached compact JWS over `payload` bytes against an Ed25519 public
 * key. Returns true iff the signature is valid and the header is well-formed
 * (alg EdDSA, b64:false, crit includes b64, empty payload segment).
 */
export function verifyDetached(jws: string, payload: Uint8Array, publicKey: Uint8Array): boolean {
  const parts = jws.split('.');
  if (parts.length !== 3 || parts[1].length !== 0) return false;
  const [protectedB64, , sigB64] = parts;
  let header: ProtectedHeader;
  try {
    header = JSON.parse(new TextDecoder().decode(base64urlDecode(protectedB64)));
  } catch {
    return false;
  }
  if (header.alg !== 'EdDSA' || header.b64 !== false || !header.crit?.includes('b64')) return false;
  try {
    return ed25519.verify(base64urlDecode(sigB64), signingInput(protectedB64, payload), publicKey);
  } catch {
    return false;
  }
}

File: packages/core-rs/src/jws.rs
//! Detached-payload JWS (compact) with Ed25519/EdDSA, per ARKY-TIM-v1 §6 and
//! RFC 7797 (b64:false). Compact form `<protected>..<signature>`; payload is the
//! JCS canonical bytes supplied separately.

use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use serde_json::json;

/// RFC 7797 signing input: ASCII(BASE64URL(protected)) || '.' || payload.
fn signing_input(protected_b64: &str, payload: &[u8]) -> Vec<u8> {
    let mut v = Vec::with_capacity(protected_b64.len() + 1 + payload.len());
    v.extend_from_slice(protected_b64.as_bytes());
    v.push(b'.');
    v.extend_from_slice(payload);
    v
}

/// Sign canonical `payload` bytes; returns compact `<protected>..<signature>`.
pub fn sign_detached(payload: &[u8], signing_key: &SigningKey, kid: Option<&str>) -> String {
    let header = match kid {
        Some(k) => json!({"alg":"EdDSA","typ":"JWS","kid":k,"b64":false,"crit":["b64"]}),
        None => json!({"alg":"EdDSA","typ":"JWS","b64":false,"crit":["b64"]}),
    };
    let protected_b64 = URL_SAFE_NO_PAD.encode(serde_json::to_vec(&header).unwrap());
    let sig: Signature = signing_key.sign(&signing_input(&protected_b64, payload));
    format!(
        "{}..{}",
        protected_b64,
        URL_SAFE_NO_PAD.encode(sig.to_bytes())
    )
}

/// Decode the protected header of a compact JWS to a serde_json Value.
pub fn decode_protected_header(jws: &str) -> Result<serde_json::Value, String> {
    let part = jws.split('.').next().ok_or("empty jws")?;
    let bytes = URL_SAFE_NO_PAD.decode(part).map_err(|e| e.to_string())?;
    serde_json::from_slice(&bytes).map_err(|e| e.to_string())
}

/// Verify a detached compact JWS over `payload` against an Ed25519 public key.
/// Returns true iff the signature is valid, the payload segment is empty, and
/// the header is well-formed (alg EdDSA, b64:false, crit includes b64).
pub fn verify_detached(jws: &str, payload: &[u8], public_key: &VerifyingKey) -> bool {
    let parts: Vec<&str> = jws.split('.').collect();
    if parts.len() != 3 || !parts[1].is_empty() {
        return false;
    }
    let (protected_b64, sig_b64) = (parts[0], parts[2]);

    let header: serde_json::Value = match decode_protected_header(jws) {
        Ok(h) => h,
        Err(_) => return false,
    };
    if header.get("alg").and_then(|v| v.as_str()) != Some("EdDSA")
        || header.get("b64").and_then(|v| v.as_bool()) != Some(false)
        || !header
            .get("crit")
            .and_then(|v| v.as_array())
            .map(|a| a.iter().any(|x| x.as_str() == Some("b64")))
            .unwrap_or(false)
    {
        return false;
    }

    let sig_bytes = match URL_SAFE_NO_PAD.decode(sig_b64) {
        Ok(b) => b,
        Err(_) => return false,
    };
    let sig = match Signature::from_slice(&sig_bytes) {
        Ok(s) => s,
        Err(_) => return false,
    };
    public_key
        .verify(&signing_input(protected_b64, payload), &sig)
        .is_ok()
}

/// Parse a 32-byte Ed25519 public key into a VerifyingKey.
pub fn verifying_key_from_bytes(bytes: &[u8]) -> Result<VerifyingKey, String> {
    let arr: [u8; 32] = bytes.try_into().map_err(|_| "key must be 32 bytes")?;
    VerifyingKey::from_bytes(&arr).map_err(|e| e.to_string())
}
